dotnet/samples/GettingStartedWithProcesses/Step04/Step04_AgentOrchestration.cs

```diff
@@ -0,0 +1,329 @@
+            .SendEventTo(new ProcessFunctionTargetBuilder(renderMessageStep, RenderMessageStep.Functions.RenderDone))
+            .StopProcess();
+
+        // Render response from primary agent
+        managerAgentStep
+            .OnEvent(AgentOrchestrationEvents.AgentResponse)
+            .SendEventTo(new ProcessFunctionTargetBuilder(renderMessageStep, RenderMessageStep.Functions.RenderMessage, parameterName: "message"));
+
+        // Request is complete
+        managerAgentStep
+            .OnEvent(CommonEvents.UserInputComplete)
+            .SendEventTo(new ProcessFunctionTargetBuilder(renderMessageStep, RenderMessageStep.Functions.RenderDone))
+            .StopProcess();
+
+        // Request more user input
+        managerAgentStep
+            .OnEvent(AgentOrchestrationEvents.AgentResponded)
+            .SendEventTo(new ProcessFunctionTargetBuilder(userInputStep));
+
+        // Delegate to inner agents
+        managerAgentStep
+            .OnEvent(AgentOrchestrationEvents.AgentWorking)
+            .SendEventTo(new ProcessFunctionTargetBuilder(managerAgentStep, ManagerAgentStep.Functions.InvokeGroup));
+
+        // Provide input to inner agents
+        managerAgentStep
+            .OnEvent(AgentOrchestrationEvents.GroupInput)
+            .SendEventTo(new ProcessFunctionTargetBuilder(agentGroupStep, parameterName: "input"));
+
+        // Render response from inner chat (for visibility)
+        agentGroupStep
+            .OnEvent(AgentOrchestrationEvents.GroupMessage)
+            .SendEventTo(new ProcessFunctionTargetBuilder(renderMessageStep, RenderMessageStep.Functions.RenderInnerMessage, parameterName: "message"));
+
+        // Provide inner response to primary agent
+        agentGroupStep
+            .OnEvent(AgentOrchestrationEvents.GroupCompleted)
+            .SendEventTo(new ProcessFunctionTargetBuilder(managerAgentStep, ManagerAgentStep.Functions.ReceiveResponse, parameterName: "response"));
+
+        KernelProcess kernelProcess = process.Build();
+
+        return kernelProcess;
+
+        void AttachErrorStep(ProcessStepBuilder step, params string[] functionNames)
+        {
+            foreach (string functionName in functionNames)
+            {
+                step
+                    .OnFunctionError(functionName)
+                    .SendEventTo(new ProcessFunctionTargetBuilder(renderMessageStep, RenderMessageStep.Functions.RenderError, "exception"))
+                    .StopProcess();
+            }
+        }
+    }
+
+    private Kernel SetupKernel(ChatHistory history)
+    {
+        IKernelBuilder builder = Kernel.CreateBuilder();
+
+        // Add Chat Completion to Kernel
+        this.AddChatCompletionToKernel(builder);
+
+        // Inject agents into service collection
+        SetupAgents(builder, builder.Build());
+        // Inject history provider into service collection
+        builder.Services.AddSingleton<IChatHistoryProvider>(new ChatHistoryProvider(history));
```

Comment:
ChatHistory provides the ChatHistoryProvider service, which contains state (the history) right? I would put it as a scoped or transient lifetime instead.

Reply:
This is a great point.  In this case the service container is the kernel itself which is scoped to this specific process...so one might consider the service container itself as _transient_.  If you look at the comment in `ChatHistoryProvider.cs`, even a provider for accessing remote history would be instantiated with transient state specific to the process instance (i.e. a `sessionId`):

```c#
CosmosDbChatHistoryProvider(CosmosClient client, string sessionId);
```

That is, the specific provider instance is bound to a particular usage.

If we were using service-wide / global dependency definitions, then the considerations would admittedly shift; although, in that case I might be tempted to put all this logic behind a _Process Provider_.